### PR TITLE
Feature/correct permission

### PR DESCRIPTION
**Pull request by MartinMeyer1**
- Fix "Transactional error"
- Fix endpoint permission update issue
- Upgrade the model element permission manager

### CODE REVIEW

File: src/main/kotlin/ch/hevs/cloudio/cloud/restapi/endpoint/permission/EndpointPermissionController.kt [hunk 1]
```diff
@@ -47,13 +47,9 @@ class EndpointPermissionController(
...
             when {
                 permission == EndpointPermission.DENY -> userEndpointPermissionRepository.deleteByUserIDAndEndpointUUID(userID, uuid)
-                permission.fulfills(EndpointPermission.WRITE) -> {
-                    val endpointPermission = userEndpointPermissionRepository.findByUserIDAndEndpointUUID(userID, uuid).orElse(UserEndpointPermission(userID, uuid, permission))
-                    endpointPermission.modelPermissions.clear()
-                    userEndpointPermissionRepository.save(endpointPermission)
-                }
                 else -> {
                     val endpointPermission = userEndpointPermissionRepository.findByUserIDAndEndpointUUID(userID, uuid).orElse(UserEndpointPermission(userID, uuid, permission))
+                    endpointPermission.permission = permission
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/restapi/endpoint/permission/EndpointPermissionController.kt:58]: We should still call `endpointPermission.modelPermissions.clear()` to get rid of model permission as they are not longer needed.

File: src/main/kotlin/ch/hevs/cloudio/cloud/restapi/endpoint/permission/EndpointPermissionController.kt [hunk 2]
```diff
@@ -63,13 +59,9 @@ class EndpointPermissionController(
...
             when {
                 permission == EndpointPermission.DENY -> userGroupEndpointPermissionRepository.deleteByUserGroupIDAndEndpointUUID(groupID, uuid)
-                permission.fulfills(EndpointPermission.WRITE) -> {
-                    val endpointPermission = userGroupEndpointPermissionRepository.findByUserGroupIDAndEndpointUUID(groupID, uuid).orElse(UserGroupEndpointPermission(groupID, uuid, permission))
-                    endpointPermission.modelPermissions.clear()
-                    userGroupEndpointPermissionRepository.save(endpointPermission)
-                }
                 else -> {
                     val endpointPermission = userGroupEndpointPermissionRepository.findByUserGroupIDAndEndpointUUID(groupID, uuid).orElse(UserGroupEndpointPermission(groupID, uuid, permission))
+                    endpointPermission.permission = permission
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/restapi/endpoint/permission/EndpointPermissionController.kt:76]: We should still call `endpointPermission.modelPermissions.clear()` to get rid of model permission as they are not longer needed.

File: src/main/kotlin/ch/hevs/cloudio/cloud/security/CloudioPermissionManager.kt [hunk 3]
```diff
@@ -111,25 +111,36 @@ class CloudioPermissionManager(
...
     fun resolveEndpointModelElementPermission(userDetails: CloudioUserDetails, modelID: ModelIdentifier): EndpointModelElementPermission {
-        var permission = EndpointModelElementPermission.DENY
+        var allPermissionsList = getAllEndpointModelElementPermissions(userDetails, modelID.endpoint)
 
-        userEndpointPermissionRepository.findByUserIDAndEndpointUUID(userDetails.id, modelID.endpoint).ifPresent {
-            it.modelPermissions[modelID.toModelPath()]?.apply {
-                permission = permission.higher(this)
-            }
+        //add the permissions with the most precise modelPath to the permissionsList
+        var temp = modelID.toString()
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/security/CloudioPermissionManager.kt:117]: IIRC the `ModelIdentifier` class should have methods to do such tasks.

**Review by cm0x4D — COMMENTED** [review]
LGTM, maybe just add the methods back that remove the model permissions as they or not needed anymore if a user has `WRITE` access to the endpoint.

File: src/main/kotlin/ch/hevs/cloudio/cloud/security/CloudioPermissionManager.kt (hunk 3 above)
Comment (MartinMeyer1) [reply src/main/kotlin/ch/hevs/cloudio/cloud/security/CloudioPermissionManager.kt:117]: I didn't find such methods

**Pull request merged**